File: cilk/chain/expertpar/outer.cpp
/*
 * outer: outer product
 *
 * input:
 *   winnow_points: a vector of (x, y) points
 *   nelts: the number of points
 *
 * output:
 *   outer_matrix: a real matrix, whose values are filled with inter-point
 *     distances
 *   outer_vector: a real vector, whose values are filled with origin-to-point
 *     distances
 */

#include <cilk/cilk.h>
#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include <algorithm>

using namespace std;

double outer_matrix[10000][10000];
double outer_vector[10000];

typedef std::pair <int, int> point; 

extern point winnow_points[10000];

double sqr(double x) {
  return x * x;
}

double distance(point a, point b) {
  return sqrt(sqr(a.first - b.first) + sqr(a.second - b.second));
}

void outer (int nelts) {
  cilk_for (int i = 0; i < nelts; ++i) {
    double nmax = 0;
    for (int j = 0; j < nelts; ++j) {
      if (i != j) {
        outer_matrix [i][j] = ::distance (winnow_points [i], winnow_points [j]);
        nmax = max (nmax, outer_matrix [i][j]);
      }      
    }
    outer_matrix [i][i] = nmax * nelts;
    outer_vector [i] = ::distance (make_pair (0,0), winnow_points [i]);
  }
}

File: erlang/chain/expertseq/outer.erl
%
% outer: outer product
%
% input:
%   points: a vector of (x, y) points
%   nelts: the number of points
%
% output:
%   matrix: a real matrix, whose values are filled with inter-point
%     distances
%   vector: a real vector, whose values are filled with origin-to-point
%     distances
%

-module(outer).
-export([outer/2]).

sqr(X) ->
  X * X.

distance({Ax, Ay}, {Bx, By}) ->
  math:sqrt(sqr(Ax - Bx) + sqr(Ay - By)).

fix_diagonal_vector(_, _, [], _, _, _) -> [];
fix_diagonal_vector(Line, Col, [Head | Tail], Point, Nelts, Nmax) ->
  if Line == Col -> [Nelts * Nmax | Tail];
    true -> [ Head | fix_diagonal_vector(
          Line, Col + 1, Tail, Point, Nelts, Nmax)]
  end.

fix_diagonal (Matrix, Points, Nelts) ->
    lists:zipwith3 
      (fun (Line, Vec, Point) ->
           fix_diagonal_vector (Line, 0, Vec, Point, Nelts, lists:max (Vec))
       end,
       lists:seq (0, length (Matrix) - 1), Matrix, Points).

outer(Nelts, Points) ->
  {fix_diagonal([ [ distance(A, B) || A <- Points] || B <- Points ], 
                Points, Nelts), 
   [distance({0, 0}, A) || A <- Points]}.
